### PR TITLE
test(benchmarks): add benchmarks for proof generation

### PR DESCRIPTION
**Pull request by ctrlc03**
# Description

Add benchmarks that compare how much time it takes to generate proofs for different  number of signups and messages. Both using WASM and Rapidsnark with c witness gen.

## Related issue(s)

Re: #1081 

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: integrationTests/ts/__benchmarks__/benchmarks.ts [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+import { Keypair, PCommand } from "maci-domainobjs";
+
+import { homedir } from "os";
+import path from "path";
+
+import { STATE_TREE_DEPTH, duration, maxValues, treeDepths, voiceCreditBalance } from "../__tests__/utils/constants";
+
+const NAME = "proof generation";
+
+// eslint-disable-next-line @typescript-eslint/require-await
```
Comment (0xmad) [inline integrationTests/ts/__benchmarks__/benchmarks.ts:13]: Why do you need to return promise if there is no await?

File: integrationTests/ts/__benchmarks__/benchmarks.ts [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+          wasmPath: path.resolve(
+            __dirname,
+            "../../../cli/zkeys/ProcessMessages_10-2-1-2_test/ProcessMessages_10-2-1-2_test_js/ProcessMessages_10-2-1-2_test.wasm",
+          ),
+          zkeyPath: path.resolve(
+            __dirname,
+            "../../../cli/zkeys/ProcessMessages_10-2-1-2_test/ProcessMessages_10-2-1-2_test.0.zkey",
+          ),
+        });
+      }
```
Comment (0xmad) [inline integrationTests/ts/__benchmarks__/benchmarks.ts:74]: This code seems is duplicated. Let's move it to a separate function and reuse for all the benchmarks.

File: integrationTests/ts/__benchmarks__/benchmarks.ts [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        maxValues,
+        treeDepths,
+        messageBatchSize,
+        coordinatorKeypair,
+      );
+      const poll = maciState.polls.get(pollId);
+
+      // signup 5 users
+      // eslint-disable-next-line @typescript-eslint/prefer-for-of
+      for (let i = 0; i < users.length; i += 1) {
```
Comment (0xmad) [inline integrationTests/ts/__benchmarks__/benchmarks.ts:37]: I think here use can use `users.forEach`. There is no await inside.

File: integrationTests/ts/__benchmarks__/benchmarks.ts [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+          pollId,
+        );
+
+        const signature = command.sign(keypair.privKey);
+
+        const ecdhKeypair = new Keypair();
+        const sharedKey = Keypair.genEcdhSharedKey(ecdhKeypair.privKey, coordinatorKeypair.pubKey);
+        const message = command.encrypt(signature, sharedKey);
+        poll?.publishMessage(message, ecdhKeypair.pubKey);
+      }
```
Comment (0xmad) [inline integrationTests/ts/__benchmarks__/benchmarks.ts:110]: Seems this one is duplicated too.

File: integrationTests/ts/__tests__/utils/constants.ts [hunk 5]
```diff
@@ -42,3 +42,16 @@ export const INT_STATE_TREE_DEPTH = 1;
...
 export const VOTE_OPTION_TREE_DEPTH = 2;
 export const MSG_BATCH_DEPTH = 1;
+
+export const voiceCreditBalance = 100n;
+export const maxValues = {
+  maxUsers: 25,
+  maxMessages: 25,
+  maxVoteOptions: 25,
+};
+export const treeDepths = {
```
Comment (0xmad) [inline integrationTests/ts/__tests__/utils/constants.ts:52]: Let's use uppercase for constant names.

File: integrationTests/ts/__benchmarks__/benchmarks.ts [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+            rapidsnarkExePath: `${homedir()}/rapidsnark/build/prover`,
+            witnessExePath: path.resolve(
+              __dirname,
+              "../../../cli/zkeys/ProcessMessages_10-2-1-2_test/ProcessMessages_10-2-1-2_test_cpp/ProcessMessages_10-2-1-2_test",
+            ),
+            zkeyPath: path.resolve(
+              __dirname,
+              "../../../cli/zkeys/ProcessMessages_10-2-1-2_test/ProcessMessages_10-2-1-2_test.0.zkey",
+            ),
+          });
```
Comment (0xmad) [inline integrationTests/ts/__benchmarks__/benchmarks.ts:351]: Is this expected that this loop is nested?

**Review by 0xmad — CHANGES_REQUESTED** [review]
@ctrlc03 thanks, just few comments.

File: integrationTests/ts/__benchmarks__/benchmarks.ts (hunk 6 above)
Comment (ctrlc03) [reply integrationTests/ts/__benchmarks__/benchmarks.ts]: nope, moved it out

File: integrationTests/ts/__benchmarks__/benchmarks.ts (hunk 2 above)
Comment (ctrlc03) [reply integrationTests/ts/__benchmarks__/benchmarks.ts:74]: sure

File: integrationTests/ts/__benchmarks__/benchmarks.ts [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+export default async function rune2e(): Promise<void> {
+  const coordinatorKeypair = new Keypair();
+
+  const messageBatchSize = 5;
+
+  const users = new Array<Keypair>(5).fill(new Keypair());
+
+  benny.suite(
+    NAME,
+
```
Comment (kittybest) [inline integrationTests/ts/__benchmarks__/benchmarks.ts:23]: why there's a blank line here? Maybe we could put some description here.

**Review by kittybest — COMMENTED** [review]
thanks!
wanna ask if we wanna gitignore the benchmark output?

**Comment by ctrlc03** [comment]
Closing now as most of the code is outdated

**Pull request closed without merging**